### PR TITLE
ARROW-5511: [Packaging] Enable Flight in Conda packages

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by pitrou** [comment]
@kszucs

**Comment by pitrou** [comment]
Does it work? I get the following error here:
```
$ conda build dev/tasks/conda-recipes/arrow-cpp/
No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.11
WARNING:conda_build.metadata:No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.11
Adding in variants from internal_defaults
INFO:conda_build.variants:Adding in variants from internal_defaults
Attempting to finalize metadata for arrow-cpp
INFO:conda_build.metadata:Attempting to finalize metadata for arrow-cpp
Traceback (most recent call last):
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/utils.py", line 1321, in _increment
    last_version = str(int(version) + 1) + suffix
ValueError: invalid literal for int() with base 10: 'none'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/antoine/miniconda3/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 456, in main
    execute(sys.argv[1:])
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 447, in execute
    verify=args.verify, variants=args.variants)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/api.py", line 208, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/build.py", line 2314, in build_tree
    notest=notest,
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/build.py", line 1324, in build
    output_metas = expand_outputs([(m, need_source_download, need_reparse_in_env)])
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/render.py", line 735, in expand_outputs
    for (output_dict, m) in _m.copy().get_output_metadata_set(permit_unsatisfiable_variants=False):
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2038, in get_output_metadata_set
    bypass_env_check=bypass_env_check)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 715, in finalize_outputs_pass
    output_d = om.get_rendered_output(metadata.name()) or {'name': metadata.name()}
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2095, in get_rendered_output
    permit_undefined_jinja=permit_undefined_jinja):
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2084, in get_rendered_outputs_section
    allow_no_other_outputs=permit_undefined_jinja)) or
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 1525, in _get_contents
    rendered = template.render(environment=env)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/home/antoine/arrow/dev/dev/tasks/conda-recipes/arrow-cpp/meta.yaml", line 13, in top-level template code
    # "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/jinja_context.py", line 324, in pin_subpackage
    exact, permit_undefined_jinja)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/jinja_context.py", line 295, in pin_subpackage_against_outputs
    max_pin))
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/utils.py", line 1351, in apply_pin_expressions
    v = _increment(v, alpha_ver)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/utils.py", line 1323, in _increment
    last_version = chr(ord(version) + 1)
TypeError: ord() expected a character, but string of length 4 found
```

**Comment by lihalite** [comment]
It's because `ARROW_VERSION` needs to be set (the CI scripts do so) `env ARROW_VERSION=0.14.0 conda build dev/tasks/conda-recipes/arrow-cpp/`

**Comment by pitrou** [comment]
Now I get another error:
```
$ ARROW_VERSION=0.14.0 conda build dev/tasks/conda-recipes/arrow-cpp/
No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.11
WARNING:conda_build.metadata:No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.11
Adding in variants from internal_defaults
INFO:conda_build.variants:Adding in variants from internal_defaults
Attempting to finalize metadata for arrow-cpp
INFO:conda_build.metadata:Attempting to finalize metadata for arrow-cpp
Collecting package metadata: ...working... done
Solving environment: ...working... failed

Leaving build/test directories:
  Work:
 /home/antoine/miniconda3/conda-bld/work 
  Test:
 /home/antoine/miniconda3/conda-bld/test_tmp 
Leaving build/test environments:
  Test:
source activate  /home/antoine/miniconda3/conda-bld/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_ 
  Build:
source activate  /home/antoine/miniconda3/conda-bld/_build_env 


Traceback (most recent call last):
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/environ.py", line 753, in get_install_actions
    actions = install_actions(prefix, index, specs, force=True)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
    return f(*args, **kwds)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/plan.py", line 473, in install_actions
    txn = solver.solve_for_transaction(prune=prune, ignore_pinned=not pinned)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/core/solve.py", line 107, in solve_for_transaction
    force_remove, force_reinstall)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/core/solve.py", line 145, in solve_for_diff
    force_remove)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/core/solve.py", line 242, in solve_final_state
    ssc = self._run_sat(ssc)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
    return f(*args, **kwds)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/core/solve.py", line 475, in _run_sat
    conflicting_specs = ssc.r.get_conflicting_specs(tuple(final_environment_specs))
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/resolve.py", line 852, in get_conflicting_specs
    reduced_index = self.get_reduced_index(specs)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
    return f(*args, **kwds)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/resolve.py", line 356, in get_reduced_index
    specs, features = self.verify_specs(specs)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda/resolve.py", line 244, in verify_specs
    raise ResolvePackageNotFound(bad_deps)
conda.exceptions.ResolvePackageNotFound: 
  - re2
  - clangdev=7

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/antoine/miniconda3/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 456, in main
    execute(sys.argv[1:])
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 447, in execute
    verify=args.verify, variants=args.variants)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/api.py", line 208, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/build.py", line 2314, in build_tree
    notest=notest,
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/build.py", line 1324, in build
    output_metas = expand_outputs([(m, need_source_download, need_reparse_in_env)])
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/render.py", line 735, in expand_outputs
    for (output_dict, m) in _m.copy().get_output_metadata_set(permit_unsatisfiable_variants=False):
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2038, in get_output_metadata_set
    bypass_env_check=bypass_env_check)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 715, in finalize_outputs_pass
    output_d = om.get_rendered_output(metadata.name()) or {'name': metadata.name()}
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2095, in get_rendered_output
    permit_undefined_jinja=permit_undefined_jinja):
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 2084, in get_rendered_outputs_section
    allow_no_other_outputs=permit_undefined_jinja)) or
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/metadata.py", line 1525, in _get_contents
    rendered = template.render(environment=env)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/home/antoine/arrow/dev/dev/tasks/conda-recipes/arrow-cpp/meta.yaml", line 46, in top-level template code
    - glog
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/jinja_context.py", line 230, in pin_compatible
    pins, _, _ = get_env_dependencies(m, 'host', m.config.variant)
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/render.py", line 143, in get_env_dependencies
    channel_urls=tuple(m.config.channel_urls))
  File "/home/antoine/miniconda3/lib/python3.7/site-packages/conda_build/environ.py", line 755, in get_install_actions
    raise DependencyNeedsBuildingError(exc, subdir=subdir)
conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform linux-64: {'clangdev=7', 're2'}
```

**Comment by lihalite** [comment]
Ah sorry, forgot conda-forge is needed, and there appears to be a config file to pin certain packages: `env ARROW_VERSION=0.14.0 conda build dev/tasks/conda-recipes/arrow-cpp/ -c conda-forge -m dev/tasks/conda-recipes/variants/osx_python3.7.yaml `

**Comment by pitrou** [comment]
Thanks :-) It finally worked.

**Comment by kszucs** [comment]
Looks good to me, but let me check via crossbow.

**Comment by pitrou** [comment]
@ursabot crossbow build -g conda

**Comment by pitrou** [comment]
@ursabot crossbow package -g conda

**Comment by pitrou** [comment]
@ursabot crossbow package -g conda

**Comment by pitrou** [comment]
@ursabot crossbow package -g conda

**Comment by pitrou** [comment]
Unfortunately the crossbow conda builds seem broken on git master. I don't think we can merge this until they're fixed :-/ @wesm @kszucs

**Comment by kszucs** [comment]
I'm working on it https://github.com/apache/arrow/pull/4649

**Comment by pitrou** [comment]
@kszucs Could you help rebasing this and fixing conflicts? Thanks :-)

**Comment by lihalite** [comment]
I just rebased this.

**Comment by kszucs** [comment]
I've just fetched it :)

**Comment by kszucs** [comment]
@ursabot crossbow package -g conda

**Comment by pitrou** [comment]
@lihalite Can you add a test that `pyarrow.flight` imports?

**Comment by lihalite** [comment]
@pitrou done.

**Comment by pitrou** [comment]
@ursabot crossbow package -g conda

**Comment by pitrou** [comment]
Not sure what happened but this build fails on uriparser: https://dev.azure.com/ursa-labs/crossbow/_build/results?buildId=123
@kszucs

**Comment by kszucs** [comment]
@ursabot crossbow package -g conda

**Comment by kszucs** [comment]
@ursabot crossbow package conda-linux-gcc-py37

**Comment by kszucs** [comment]
@ursabot crossbow package conda-linux-gcc-py37

**Comment by kszucs** [comment]
@ursabot crossbow package conda-linux-gcc-py37

File: dev/tasks/conda-recipes/arrow-cpp/build.sh [hunk 1]
```diff
@@ -19,12 +19,14 @@ cmake \
     -DCMAKE_INSTALL_LIBDIR=$PREFIX/lib \
     -DLLVM_TOOLS_BINARY_DIR=$PREFIX/bin \
     -DARROW_DEPENDENCY_SOURCE=SYSTEM \
+    -Duriparser_SOURCE=BUNDLED \
```
Comment (xhochy) [inline dev/tasks/conda-recipes/arrow-cpp/build.sh]: Note that we need to package `uriparser` before we can release this on conda-forge; there bundling of thirdparties is not really accepted.

Comment (kszucs) [reply dev/tasks/conda-recipes/arrow-cpp/build.sh]: Thanks for the update, didn't know that.

Comment (pitrou) [reply dev/tasks/conda-recipes/arrow-cpp/build.sh]: Even for small private dependencies?

**Comment by kszucs** [comment]
@ursabot crossbow package -g conda

**Comment by kszucs** [comment]
@ursabot crossbow package conda-osx-clang-py27 conda-osx-clang-py36

**Comment by kszucs** [comment]
Conda probably doesn't move the submodules to the testing directory, skipping the tests instead.

**Comment by kszucs** [comment]
@ursabot crossbow package -g conda

File: dev/tasks/conda-recipes/arrow-cpp/build.sh (hunk 1 above)
Comment (wesm) [reply dev/tasks/conda-recipes/arrow-cpp/build.sh]: I think we can temporarily statically-link uriparser, we had a bunch of stuff statically linked on conda-forge for a long time before the CMake refactor

**Comment by wesm** [comment]
Ugly mess of conda-build warnings in the Windows build

https://gist.github.com/wesm/fdbdcfb0882ecba6d27d563566c90532

I'm looking at the Windows build script

**Comment by wesm** [comment]
@ursabot crossbow package conda-win-vs2015-py36 conda-win-vs2015-py37

**Comment by wesm** [comment]
@ursabot crossbow package conda-win-vs2015-py36 conda-win-vs2015-py37

**Review by wesm — APPROVED** [review]
+1. The Windows builds are passing now

**Comment by wesm** [comment]
Crossbow builds https://github.com/wesm/crossbow/branches/all?utf8=%E2%9C%93&query=build-27

**Comment by kszucs** [comment]
@wesm thanks for fixing the windows builds. When use use [skip ci] pattern in the commit message, then ursabot will skip executing that particular commit. It is because we proxy the request to the standard PR building mechanism of buildbot.

**Pull request closed without merging**